File: java/src/SudokuSolver/Controller.java
package SudokuSolver;

import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.scene.control.Button;
import javafx.scene.control.TextField;

public class Controller {
    private View view;
    private DancingLinks dlx;

    public Controller(View view) {
        this.view = view;
        dlx = new DancingLinks(new ExactCoverMatrix("ExactCoverMatrix.txt"));
        view.getButton().setOnAction(new ButtonSolveAction());
    }

    class ButtonSolveAction implements EventHandler<ActionEvent> {
        @Override
        public void handle(ActionEvent actionEvent) {
            long startTime = System.nanoTime();

            TextField[][] cells = view.getCells();
            for (int i = 0; i < 9; ++i) {
                for (int j = 0; j < 9; ++j) {
                    String text = cells[i][j].getText();
                    if (text.length() > 0) {
                        dlx.setAssignment(i, j, Integer.parseInt(text)); // Assign the user-chosen cells
                    }
                }
            }

            int[][] solution = dlx.solve();
            long elapsedTime = System.nanoTime() - startTime; // Algorithm running time in nanoseconds

            view.setAllCells(solution); // Solve using dancing links and display
            if (solution != null) {
                view.displayMessage(String.format("Solved in %.9f seconds", (double) elapsedTime / 1_000_000_000));
            } else {
                view.displayMessage("No solution exists");
            }


            Button b = view.getButton();
            b.setText("R E S E T");
            b.setOnAction(new ButtonResetAction());
        }
    }

    class ButtonResetAction implements EventHandler<ActionEvent> {
        @Override
        public void handle(ActionEvent actionEvent) {
            view.resetCells();                      // Reset sudoku board
            Button b = view.getButton();
            b.setText("S O L V E");
            b.setOnAction(new ButtonSolveAction());
        }
    }
}

File: typescript/src/SudokuSolver/Controller.ts
import { View } from './View';
import { DancingLinks } from './DancingLinks';
import { ExactCoverMatrix } from './ExactCoverMatrix';

export class Controller {
    private view: View;
    private dlx: DancingLinks;
    private isSolved: boolean = false;

    constructor(view: View) {
        this.view = view;
        this.dlx = new DancingLinks(new ExactCoverMatrix());
        this.view.getButton().addEventListener('click', () => this.handleButtonClick());
    }

    private handleButtonClick(): void {
        if (this.isSolved) {
            this.handleReset();
        } else {
            this.handleSolve();
        }
    }

    private handleSolve(): void {
        const startTime = performance.now();

        const cells = this.view.getCells();
        for (let i = 0; i < 9; i++) {
            for (let j = 0; j < 9; j++) {
                const text = cells[i][j].value;
                if (text.length > 0) {
                    this.dlx.setAssignment(i, j, parseInt(text, 10));
                }
            }
        }

        const solution = this.dlx.solve();
        const elapsedTime = performance.now() - startTime;

        this.view.setAllCells(solution);

        if (solution !== null) {
            this.view.displayMessage(`Solved in ${(elapsedTime / 1000).toFixed(9)} seconds`);
        } else {
            this.view.displayMessage('No solution exists');
        }

        const button = this.view.getButton();
        button.textContent = 'R E S E T';
        this.isSolved = true;
    }

    private handleReset(): void {
        // Create new DLX instance for fresh solve
        this.dlx = new DancingLinks(new ExactCoverMatrix());

        this.view.resetCells();
        const button = this.view.getButton();
        button.textContent = 'S O L V E';
        this.isSolved = false;
    }
}
